File: allsportsapi/Examples/list_all_leagues.js
/**
 * List every league of a sport — a runnable version of the FAQ recipe
 * "How do I get all leagues of a sport?" (see ../FAQ.md).
 *
 * Strategy (top-down tree walk):
 *   1. List all categories:          /api/tournament/categories
 *   2. For each category, list its
 *      unique tournaments (leagues):  /api/tournament/all/category/{id}
 *
 * This example targets football (host footapi7). For other sports, prefix the path
 * with the sport, e.g. /api/basketball/tournament/categories, and set RAPIDAPI_HOST
 * (and SPORT_PREFIX) accordingly.
 *
 * Run (Node 18+):
 *     export RAPIDAPI_KEY="your_key_here"
 *     node list_all_leagues.js
 *
 * NOTE: this makes one request per category, so it can be many calls. Categories and
 * leagues change slowly — cache the result (hours/days) instead of re-walking often.
 */

const HOST = process.env.RAPIDAPI_HOST || 'footapi7.p.rapidapi.com';
const BASE_URL = `https://${HOST}`;
const API_KEY = process.env.RAPIDAPI_KEY;

// Path prefix: "" for football, "/basketball", "/tennis", … for other sports.
const SPORT_PREFIX = process.env.SPORT_PREFIX || '';

const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

async function get(path) {
  if (!API_KEY) {
    console.error('Set RAPIDAPI_KEY in your environment first.');
    process.exit(1);
  }
  const headers = { 'X-RapidAPI-Key': API_KEY, 'X-RapidAPI-Host': HOST };
  const res = await fetch(`${BASE_URL}${path}`, { headers });
  if (res.status === 204) return null;
  if (!res.ok) throw new Error(`HTTP ${res.status} for ${path}`);
  return res.json();
}

async function listCategories() {
  const data = await get(`/api${SPORT_PREFIX}/tournament/categories`);
  return data?.categories ?? [];
}

async function listLeaguesInCategory(categoryId) {
  const data = await get(`/api${SPORT_PREFIX}/tournament/all/category/${categoryId}`);
  return data?.uniqueTournaments ?? [];
}

async function main() {
  const categories = await listCategories();
  console.log(`${categories.length} categories\n`);

  let totalLeagues = 0;
  for (const category of categories) {
    const leagues = await listLeaguesInCategory(category.id);
    totalLeagues += leagues.length;
    console.log(`${category.name} (${category.alpha2 ?? '—'}) — ${leagues.length} leagues`);
    for (const league of leagues) {
      console.log(`    [${league.id}] ${league.name}`);
    }
    await sleep(200); // be gentle on your rate limit
  }

  console.log(`\nTotal: ${totalLeagues} leagues across ${categories.length} categories`);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: allsportsapi/Examples/list_all_leagues.py
"""
List every league of a sport — a runnable version of the FAQ recipe
"How do I get all leagues of a sport?" (see ../FAQ.md).

Strategy (top-down tree walk):
    1. List all categories:            /api/tournament/categories
    2. For each category, list its
       unique tournaments (leagues):   /api/tournament/all/category/{id}

This example targets football (host footapi7). For other sports, prefix the path
with the sport, e.g. /api/basketball/tournament/categories and
/api/basketball/tournament/all/category/{id}, and set RAPIDAPI_HOST accordingly.

Run:
    export RAPIDAPI_KEY="your_key_here"
    pip install requests
    python list_all_leagues.py

NOTE: this makes one request per category, so it can be many calls. Categories and
leagues change slowly — cache the result (hours/days) instead of re-walking often.
"""

import os
import sys
import time

import requests

HOST = os.environ.get("RAPIDAPI_HOST", "footapi7.p.rapidapi.com")
BASE_URL = f"https://{HOST}"
API_KEY = os.environ.get("RAPIDAPI_KEY")

# Path prefix: "" for football, "/basketball", "/tennis", … for other sports.
SPORT_PREFIX = os.environ.get("SPORT_PREFIX", "")


def get(path, *, timeout=15):
    if not API_KEY:
        sys.exit("Set RAPIDAPI_KEY in your environment first.")
    headers = {"X-RapidAPI-Key": API_KEY, "X-RapidAPI-Host": HOST}
    resp = requests.get(f"{BASE_URL}{path}", headers=headers, timeout=timeout)
    if resp.status_code == 204:
        return None
    resp.raise_for_status()
    return resp.json()


def list_categories():
    data = get(f"/api{SPORT_PREFIX}/tournament/categories")
    return (data or {}).get("categories", [])


def list_leagues_in_category(category_id):
    data = get(f"/api{SPORT_PREFIX}/tournament/all/category/{category_id}")
    return (data or {}).get("uniqueTournaments", [])


def main():
    categories = list_categories()
    print(f"{len(categories)} categories\n")

    total_leagues = 0
    for category in categories:
        leagues = list_leagues_in_category(category["id"])
        total_leagues += len(leagues)
        print(f"{category['name']} ({category.get('alpha2', '—')}) — {len(leagues)} leagues")
        for league in leagues:
            print(f"    [{league['id']}] {league['name']}")
        time.sleep(0.2)  # be gentle on your rate limit

    print(f"\nTotal: {total_leagues} leagues across {len(categories)} categories")


if __name__ == "__main__":
    main()
